src/Input.test.jsx

```diff
@@ -0,0 +1,113 @@
+import React from 'react';
+
+import { render, fireEvent } from '@testing-library/react';
+
+import Input from './Input';
+
+describe('Input', () => {
+  const setup = ({
+    value,
+    handleChange = jest.fn(),
+    handleClick = jest.fn(),
+  }) => {
+    const utils = render(
+      <Input
+        value={value}
+        onClick={handleClick}
+        onChange={handleChange}
+      />,
+    );
+
+    return { ...utils };
+  };
```

Comment:
setup을 통해 각기 다른 테스트 케이스마다 props들을 세팅해주어 다양한 테스트를 작성하신것 같네요! 좋습니다👍
다만 조금 복잡한 것 같아요~ handleChange와 handleClick의 경우, 최초 한번만 선언해도 될 것 같은데 어떤가요?

```suggestion
  const handleChange = jest.fn();
  const handleClick = jest.fn();

  const renderInput = render((
    <Input
      value={value}
      onClick={handleClick}
      onChange={handleChange}
    />
  ));
```

그리고 setup대신에 각 it함수 내에서 renderInput을 호출하는 방식으로 처리하면 어떨까요?

Reply:
음... 조언해주신 부분이 어떤 식으로 적용해야 되는건지 잘 모르겠습니다 ㅠㅠ

```javascript
describe('Input', () => {
  const handleChange = jest.fn();
  const handleClick = jest.fn();
  const placeholderText = '할 일을 입력해 주세요';
  const buttonText = '추가';
  const value = '';
  const renderInput = render((
    <Input
      value={value}
      onClick={handleClick}
      onChange={handleChange}
    />
  ));

  context('with value', () => {
    it('check elements', () => {
      renderInput.getByPlaceholderText(placeholderText);
      renderInput.getByText(buttonText);
    });

    it('click button', () => {
      const button = renderInput.getByText(buttonText);

      expect(handleClick).not.toBeCalled();

      fireEvent.click(button);

      expect(handleClick).toBeCalledTimes(1);
    });
  });
});
```

이런 방식을 말씀하신건지

```javascript
describe('Input', () => {
  const handleChange = jest.fn();
  const handleClick = jest.fn();
  const placeholderText = '할 일을 입력해 주세요';
  const buttonText = '추가';

  context('with value', () => {
    const value = '';
    const renderInput = render((
      <Input
        value={value}
        onClick={handleClick}
        onChange={handleChange}
      />
    ));

    it('check elements', () => {
      renderInput.getByPlaceholderText(placeholderText);
      renderInput.getByText(buttonText);
    });

    it('click button', () => {
      const button = renderInput.getByText(buttonText);

      expect(handleClick).not.toBeCalled();

      fireEvent.click(button);

      expect(handleClick).toBeCalledTimes(1);
    });
  });
});
```

아니면 이런 방식을 말씀하신건지
아니면 아래 방법인지

```javascript
import React from 'react';

import { render, fireEvent } from '@testing-library/react';

import Input from './Input';

describe('Input', () => {
  const handleChange = jest.fn();
  const handleClick = jest.fn();
  const placeholderText = '할 일을 입력해 주세요';
  const buttonText = '추가';

  context('with value', () => {
    const value = '';

    it('check elements', () => {
      const renderInput = render((
        <Input
          value={value}
          onClick={handleClick}
          onChange={handleChange}
        />
      ));

      renderInput.getByPlaceholderText(placeholderText);
      renderInput.getByText(buttonText);
    });

    it('click button', () => {
      const renderInput = render((
        <Input
          value={value}
          onClick={handleClick}
          onChange={handleChange}
        />
      ));
      const button = renderInput.getByText(buttonText);

      expect(handleClick).not.toBeCalled();

      fireEvent.click(button);

      expect(handleClick).toBeCalledTimes(1);
    });
  });
});

```

이래저래 고민해보면서 작성해보았는데, 첫번째, 두번째 방법은 에러가 나서 실패하고
세번째 방법은 단순화 시키려면 setup() 같은 방식을 사용할 수 밖에 없게 되서...
결론적으로는 조언해주신 방향을 전혀 이해하지 못했습니다 ㅠㅠ

Reply:
제가 설명이 부족했던 것 같네요~ setup()을 사용하는 방식과 크게 다르지 않습니다.

```
describe('Input', () => {
  const handleChange = jest.fn();
  const handleClick = jest.fn();
  const placeholderText = '할 일을 입력해 주세요';
  const buttonText = '추가';

  context('with value', () => {
    const value = '';

    it('check elements', () => {
      const renderInput = render((
        <Input
          value={value}
          onClick={handleClick}
          onChange={handleChange}
        />
      ));

      renderInput.getByPlaceholderText(placeholderText);
      renderInput.getByText(buttonText);
    });

    it('click button', () => {
      const renderInput = 
      const button = renderInput.getByText(buttonText);

      expect(handleClick).not.toBeCalled();

      fireEvent.click(button);

      expect(handleClick).toBeCalledTimes(1);
    });
  });
});
```

Reply:
context가 한개일 땐 괜찮은데
2가지 이상으로 다른 상태에서 동일한 테스트를 진행하려니 문제가 생깁니다 ㅠㅠ
onClickTest와 onChangeTest를 풀어서 써도 동일하게 문제가 발생합니다

일단 문제가 생기는 상태의 코드를 푸시해뒀습니다 ㅠㅠ